crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1023,18 +1031,104 @@ impl<'db> TypeInferenceBuilder<'db> {
         with_item: &ast::WithItem,
         definition: Definition<'db>,
     ) {
-        let expression = self.index.expression(&with_item.context_expr);
-        let result = infer_expression_types(self.db, expression);
-        self.extend(result);
+        let context_expr = self.index.expression(&with_item.context_expr);
+        self.extend(infer_expression_types(self.db, context_expr));
 
-        // TODO(dhruvmanila): The correct type inference here is the return type of the __enter__
-        // method of the context manager.
-        let context_expr_ty = self.expression_ty(&with_item.context_expr);
+        let target_ty = self.infer_context_expression(
+            &with_item.context_expr,
+            self.expression_ty(&with_item.context_expr),
+        );
 
         self.types
             .expressions
-            .insert(target.scoped_ast_id(self.db, self.scope()), context_expr_ty);
-        self.add_binding(target.into(), definition, context_expr_ty);
+            .insert(target.scoped_ast_id(self.db, self.scope()), target_ty);
+        self.add_binding(target.into(), definition, target_ty);
+    }
+
+    /// Infers the type of a context expression (`with expr`) and returns the target's type
+    ///
+    /// Returns [`Type::Unknown`] if the context expression doesn't implement the context manager protocol.
+    ///
+    /// ## Terminology
+    /// See [PEP343](https://peps.python.org/pep-0343/#standard-terminology).
+    fn infer_context_expression(
+        &mut self,
+        context_expression: &ast::Expr,
+        context_expression_ty: Type<'db>,
+    ) -> Type<'db> {
+        let context_manager_ty = context_expression_ty.to_meta_type(self.db);
+
+        let enter_ty = context_manager_ty.member(self.db, "__enter__");
+        let exit_ty = context_manager_ty.member(self.db, "__exit__");
+
+        // TODO: The checks here should be simplified to checking if context manager implements the `contextlib.AbstractContextManager` protocol.
+        if enter_ty.is_unbound() && exit_ty.is_unbound() {
+            self.add_diagnostic(
+                context_expression.into(),
+                "invalid-context-manager",
+                format_args!(
+                    "Object of type {} cannot be used with `with` because it doesn't implement `__enter__` and `__exit__`",
```

Comment:
The detail in these error messages is fantastic! Long-term, I'm not sure we'll want all this information in one long sentence like this, but I think this is fine for now

Reply:
Agree